Add audit trail

stevehook opened the pull request:
### Context

We want to have an audit trail of changes made via the various interfaces to application data (`ApplicationForm`, `ApplicationChoice` and friends) to assist with debugging issues in development and support in production. We expect the audit trail to tell us who made what changes to a given application and when.

### Changes proposed in this pull request

This PR is just about the backend changes needed to start recording auditing information in our database.

* Adds and configures the `audited` gem (also used in the Find service)
* All create/update/delete operations on `ApplicationForm` and `ApplicationChoice` are recorded.
* `ApplicationChoice` audit records are _associated_ with their parent `ApplicationForm` so that we can get a unified view of the audit records for a whole application.
* As further model attributes/classes are added they will also be configured to audit changes and be associated with the parent `ApplicationForm`
* Audit records triggered by the candidate interface are associated with a `Candidate` user.
* Audit records triggered by the vendor API are associated with a (new) `VendorApiUser` that encapsulates what we know about the end user of the API as declared in vendor attribution metadata.

![domain-model](https://user-images.githubusercontent.com/450843/67499831-9acb4800-f679-11e9-9a84-0fdae45d76e0.png)

Out of scope here but to be covered by a follow-up PR are changes to the Support UI to make audit trail data visible to developers and support staff.

### Guidance to review

Things to look out for:

* Naming - we've introduced a new `VendorApiUser` model among other things.
* Specs - auditing is a kind of cross-cutting concern. I wanted to test it at the request spec level to assert that audit records and created and that they are linked to the correct 'user' (which varies depending on the type of request). It didn't seem right to test this kind of stuff at the system spec level because it's about stuff that is hidden from the front-end.
* I've added a spec or two for each type of request that we have some kind of authenticated user for at the moment - candidate and vendor API. I'd like to add more specs at this level to all actions in these areas that trigger an audit record but I don't want to swamp this PR and wanted to check what others think first. Thoughts welcome!

### Link to Trello card

[1119 - Add audit trail for all updates](https://trello.com/c/7ubeCZJL/1119-add-audit-trail-for-all-updates-%F0%9F%8F%88-kickoff)

duncanjbrown commented:
This seems to work well. I created an application, filled in my personal details, then made an offer via the API. Made a summary, below, using this:

```
 JSON.pretty_generate(Audited::Audit.all.map {|a| [a.created_at, a.action + ' ' + a.auditable_type, a.audited_changes, a.user.try(:email_address) || a.user.try(:email)] })
```

An interesting bit of grit here is that VendorApiUser and Candidate don't have the same interface so I needed to map between them when making this synopsis. (Not sure that we'll need to mix these things up in future, but in principle having a consistent name for such fields seems sensible)


```
[
  [
    "2019-10-24 13:42:57 UTC",
    "create ApplicationChoice",
    {
      "offer": null,
      "status": "unsubmitted",
      "course_option_id": 1,
      "rejection_reason": null,
      "personal_statement": null,
      "application_form_id": 1
    },
    "duncan@duncanjbrown.com"
  ],
  [
    "2019-10-24 13:42:57 UTC",
    "create ApplicationForm",
    {
      "last_name": null,
      "first_name": null,
      "candidate_id": 1,
      "submitted_at": null,
      "date_of_birth": null,
      "first_nationality": null,
      "second_nationality": null,
      "further_information": null,
      "english_main_language": null,
      "other_language_details": null,
      "english_language_details": null,
      "further_information_details": null
    },
    "duncan@duncanjbrown.com"
  ],
  [
    "2019-10-24 13:43:21 UTC",
    "update ApplicationForm",
    {
      "last_name": [
        null,
        "Briggs"
      ],
      "first_name": [
        null,
        "Lamar"
      ],
      "date_of_birth": [
        null,
        "2011-06-20"
      ],
      "first_nationality": [
        null,
        "Moroccan"
      ],
      "second_nationality": [
        null,
        ""
      ],
      "english_main_language": [
        null,
        true
      ],
      "other_language_details": [
        null,
        ""
      ],
      "english_language_details": [
        null,
        "I also speak Latin"
      ]
    },
    "duncan@duncanjbrown.com"
  ],
  [
    "2019-10-24 13:43:41 UTC",
    "update ApplicationForm",
    {
      "further_information": [
        null,
        true
      ],
      "further_information_details": [
        null,
        "Oh I am good at teaching"
      ]
    },
    "duncan@duncanjbrown.com"
  ],
  [
    "2019-10-24 13:43:42 UTC",
    "update ApplicationChoice",
    {
      "status": [
        "unsubmitted",
        "application_complete"
      ]
    },
    "duncan@duncanjbrown.com"
  ],
  [
    "2019-10-24 13:43:42 UTC",
    "update ApplicationForm",
    {
      "submitted_at": [
        null,
        "2019-10-24T13:43:42.024Z"
      ]
    },
    "duncan@duncanjbrown.com"
  ],
  [
    "2019-10-24 13:45:22 UTC",
    "update ApplicationChoice",
    {
      "status": [
        "application_complete",
        "conditional_offer"
      ]
    },
    "pat@greendale.co.uk"
  ],
  [
    "2019-10-24 13:45:22 UTC",
    "update ApplicationChoice",
    {
      "offer": [
        null,
        {
          "conditions": [
            "Confirmation of identity"
          ]
        }
      ]
    },
    "pat@greendale.co.uk"
  ]
]
```

mnacos commented:
Just a couple of non-blocking comments from me:

- VendorApiUser has not made it into the data model diagram, and
- I had a similar experience when writing specs for basic auth, because this was a cross-cutting concern, too, the way we scoped it. I went the other way and created request specs but placed mine in system/ because the tests there are closer to feature specs. This suggests to me we currently lack the a proper location for cross-cutting rspecs, which could arguably go into either requests or system.

duncanjbrown commented:
> thanks @duncanjbrown
> 
> I guess we can make the interfaces for Candidate and VendorApiUser consistent. That's an oversight on my part.
> 
> How about we:
> 
> * rename `VendorApiUser#email` to `email_address`
> * alias `Candidate#id` as `user_id`
> * add a `Candidate#full_name` method that pulls the first and last names from `current_application`?

This sounds good. For `Candidate#full_name`, I guess it would look at `application_forms.last`?

stevehook commented:
@duncanjbrown yes `application_forms.last` makes more sense than `current_application` I think.

duncanjbrown commented:
Thinking about it, do we need `user_id` on the `Candidate`? The `user_id` on `VendorApiUser` is just a bit of foreign metadata.

stevehook commented:
It feels a bit odd adding a `Candidate#full_name` method come to think of it. Up to now we've kept this kind of logic in presenters, there is already one presenter that implements this. I'll add it for now but if anybody has any better ideas about where to put it please shout.

stevehook commented:
> Thinking about it, do we need `user_id` on the `Candidate`? The `user_id` on `VendorApiUser` is just a bit of foreign metadata.
fair point, `VendorApiUser#user_id` isn't going to be much use to us.

stevehook commented:
After discussion with @duncanjbrown I have:

- Moved the request specs that I added to test the auditing cross cutting concern into their own spec files so that they are not muddled up with the specs that test core application logic.
- Renamed `VendorApiUser#email` to `email_address` to be a bit consistent with `Candidate`.

stevehook commented:
> Just a couple of non-blocking comments from me:
> 
> * VendorApiUser has not made it into the data model diagram, and
> * I had a similar experience when writing specs for basic auth, because this was a cross-cutting concern, too, the way we scoped it. I went the other way and created request specs but placed mine in system/ because the tests there are closer to feature specs. This suggests to me we currently lack the a proper location for cross-cutting rspecs, which could arguably go into either requests or system.

Thanks @mnacos , I've corrected the ERD diagram and hopefully the changes I made to separate out the cross-cutting request specs will help. In the longer term it might be better to have these as system specs but we will need to implement some kind of front-end for it first.

duncanjbrown commented on app/controllers/vendor_api/vendor_api_controller.rb:
```diff
@@ -8,6 +8,17 @@ class VendorApiController < ActionController::API
+      return nil unless @metadata.present?
+
+      @audit_user ||= VendorApiUser.find_or_create_by(
+        email_address: @metadata.attribution.email,
```
This will recreate a user if their name changes in the vendor system. Do we want to find_or_create on the basis of only `user_id` and `vendor_api_token_id`?

stevehook replied:
Happy to change it but I'm guessing that the vendor API token can change too, though not often. Also we need to consider whether we want the information we capture to reflect the state of the user at the time or what they are now (or at least were when they last accessed the API)

duncanjbrown commented on app/controllers/vendor_api/vendor_api_controller.rb:
```diff
@@ -8,6 +8,17 @@ class VendorApiController < ActionController::API
+      @audit_user ||= VendorApiUser.find_or_create_by(
+        email_address: @metadata.attribution.email,
+        full_name: @metadata.attribution.full_name,
+        user_id: @metadata.attribution.user_id,
```
Should we namespace this a bit? `vendor_user_id`, maybe? just to disambiguate from the `user_id` in the audits table
